PIVX-Labs/MyPIVXWallet
# [Mempool] Mempool V2

## Abstract

This PR completely rewrites the mempool system. 

First of all the concepts of `Transaction`, `CTxIn`, `CTxOut` and `COutpoint` have been formally defined with classes. 
The wallet can now iterate through the outputs of a transaction and get the output type (only P2Cs and P2PKH for the moment) and check if it has the private key to spend it.
For the mempool the old UTXO array has been removed in place of two new maps:

- `txmap: map txid -> Transaction`  which stores a subset of transaction that involves the user
- `spent: multimap txid -> COutpoint` which stores all the `vout` that are spent.

Sync happens in two steps:

1. When the wallet is first opened we fetch from explorer all our UTXOs and build `txmap` and `spent`
2. Then all new transactions that are generated while the user is using MPW are fetched with the event `recent_txs`. In particular when a new `transaction` is received it is added to `txmap` and all its `CTxIn` are put in the `spent` multimap.

Getting UTXO from `txmap` is done in this way: we check every `CTxOut` of every `Transcation` if:

1.  The outpoint of `CTxOut` is not yet in the `spent` multimap
2. We have the private key to spend `CTxOut` 

Then `CTxOut`  is a valid UTXO,

The main advantage of this system is that we save a lot of network requests: indeed the full UTXOs list is fetched only once when the wallet is first opened ( and not at every new block).  The main disadvantage  instead is that the wallet will be a bit slower (especially when loading `txmap` and `spent` for the first time) .

## Testing

- Check that transactions (send, receive, etc) are working fine, no balance duplication, etc.
- Open two MPW instances, either in different browsers or devices, Send and Receive; ensure there's no duplication between them.
- Check the speed of balance updates, both Sending and Receiving, is it too slow?
- Check that you can create rapid consecutive transactions, and verify the speed is good enough.
- Check that PIVX promos are working, Creating and Redeeming.
---

## For future
For reviewers: yes I agree with you that the new mempool looks overkill and complex with respect to the old "I trivially save unspent utxos". 
However take in consideration that this has been designed to be much more than a simple mempool: we have a system that given a (sufficient) set of transactions can "do everything" without having to rely on an explorer.

A list of future features that can be built on this are:

- The activity can simply be built from the new mempool instead of asking each time the explorer (saving tons of bandwidth)
- The network system can be hugely simplified: We only need to ask for transactions (first time sync + a new request at each block to see if there are new txs), the wallet is able to handle everything else on its own
- Those transactions can be then all stored on DB (which would be another huge improvement on bandwidth saved)
